src/app/backend/handler/attach.go

```diff
@@ -0,0 +1,117 @@
+package handler
+
+import (
+	"encoding/json"
+        "fmt"
+        "net/http"
+	"log"
+
+	"k8s.io/kubernetes/pkg/client/unversioned/remotecommand"
+        "gopkg.in/igm/sockjs-go.v2/sockjs"
+)
+
+// Implements PtyHandler
+type AttachSession struct {
+  id string
+  bound chan error
+  sockJSSession sockjs.Session
+  sizeChan chan remotecommand.TerminalSize
+}
+
+func (t AttachSession) Next() *remotecommand.TerminalSize {
+	select {
+	case size := <-t.sizeChan:
+		return &size
+	}
+	return nil
+}
+
+func (t AttachSession) Read(p []byte) (int, error) {
+        m, err := t.sockJSSession.Recv()
+        if err != nil {
+          return 0, err
+        }
+
+	var msg struct {
+		Op, Data   string
+		Rows, Cols uint16
+	}
+	if err := json.Unmarshal([]byte(m), &msg); err != nil {
+          return 0, err
+	}
+
+	switch msg.Op {
+	case "stdin":
+		return copy(p, msg.Data), nil
+	case "resize":
+		t.sizeChan <- remotecommand.TerminalSize{msg.Cols, msg.Rows}
+		return 0, nil
+	default:
+		return 0, fmt.Errorf("unknown message type '%s'", msg.Op)
+	}
+}
+
+func (t AttachSession) Write(p []byte) (int, error) {
+	type Message struct {
+		Op, Data string
+	}
+	msg, err := json.Marshal(Message{"stdout", string(p)})
+	if err != nil {
+		return 0, err
+	}
+
+        if err = t.sockJSSession.Send(string(msg)); err != nil {
+		return 0, err
+	}
+	return len(p), nil
+}
+
+func (t AttachSession) Toast(p string) (error) {
+	type Message struct {
+		Op, Data string
+	}
+	msg, err := json.Marshal(Message{"toast", p})
+	if err != nil {
+		return err
+	}
+
+        if err = t.sockJSSession.Send(string(msg)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t AttachSession) Close(status uint32, reason string) {
+  t.sockJSSession.Close(status, reason)
+}
+
+// FIXME: this structure needs locking
+var attachSessions = make(map[string]AttachSession)
+
+func handleAttachSession(session sockjs.Session) {
+  if buf, err := session.Recv(); err == nil {
```

Comment:
I'd reverse the logic here to decrease complexity.

```
if ...; err != nil {
log(err)
return
}
```

Also in case of any error we can probably just retun immediately, then last part does not have to be in else block.